File: core/src/main/java/org/predicode/predicator/RulePattern.java
package org.predicode.predicator;

import org.predicode.predicator.predicates.Predicate;
import org.predicode.predicator.terms.PlainTerm;

import javax.annotation.Nonnull;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static org.predicode.predicator.grammar.TermPrinter.printTerms;


/**
 * Resolution rule match pattern.
 */
public final class RulePattern {

    @Nonnull
    private final List<? extends PlainTerm> terms;

    /**
     * Constructs rule pattern.
     *
     * @param terms a list of terms this pattern consists of.
     */
    public RulePattern(@Nonnull List<? extends PlainTerm> terms) {
        this.terms = Collections.unmodifiableList(terms);
    }

    /**
     * Constructs rule pattern out of terms array.
     *
     * @param terms array of terms this pattern consists of.
     */
    public RulePattern(@Nonnull PlainTerm... terms) {
        this(Arrays.asList(terms));
    }

    /**
     * A list of terms this pattern consists of.
     *
     * @return readonly list of terms.
     */
    @Nonnull
    public final List<? extends PlainTerm> getTerms() {
        return this.terms;
    }

    /**
     * Attempts to match the given predicate call against this pattern.
     *
     * <p>This method is called for the {@link Rule#getCondition() rule condition} with target call as argument.
     *
     * @param call a predicate call to match.
     * @param knowns known resolutions.
     *
     * @return updated knowns if the call matches this pattern, or empty optional otherwise.
     */
    @Nonnull
    public Optional<Knowns> match(@Nonnull Predicate.Call call, @Nonnull Knowns knowns) {
        if (!call.isFinite()) {
            // TODO implement infinite predicate call matching against open rule patterns.
            return Optional.empty();
        }
        if (getTerms().size() != call.length()) {
            return Optional.empty(); // Wrong call length
        }

        return call.prefix(getTerms().size()).flatMap(prefix -> {

            @Nonnull
            Knowns result = knowns.startMatching();
            int index = 0;

            for (PlainTerm term : getTerms()) {

                final Optional<Knowns> match = term.match(prefix.getTerms().get(index), result);

                if (!match.isPresent()) {
                    return Optional.empty();
                }

                result = match.get();
                ++index;
            }

            return Optional.of(result);
        });
    }

    /**
     * Creates a resolution rule with this pattern as its {@link Rule#getCondition() condition}.
     *
     * @param predicate predicate the constructed rule resolves to if this pattern matches.
     */
    @Nonnull
    public final Rule rule(@Nonnull Predicate predicate) {
        return new Rule(this, predicate);
    }

    /**
     * Creates a fact with this pattern as its {@link Rule#getCondition() condition}.
     */
    @Nonnull
    public final Rule fact() {
        return rule(Predicate.TRUE);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        final RulePattern that = (RulePattern) o;

        return this.terms.equals(that.terms);
    }

    @Override
    public int hashCode() {
        return this.terms.hashCode();
    }

    @Override
    public String toString() {
        return printTerms(getTerms());
    }

}

File: core/src/main/kotlin/org/predicode/predicator/RulePattern.kt
package org.predicode.predicator

import org.predicode.predicator.grammar.printTerms
import reactor.core.publisher.Flux

/**
 * Resolution rule match pattern.
 *
 * @constructor constructs rule pattern.
 *
 * @property terms list of terms this pattern consists of.
 */
class RulePattern(val terms: List<PlainTerm>) : Predicate {

    /**
     * Constructs new rule pattern out of terms array.
     *
     * @param terms array of terms this pattern consists of.
     */
    constructor(vararg terms: PlainTerm) : this(terms.asList())

    /**
     * Attempts to match against another pattern.
     *
     * This method is called for the [rule condition][Rule.condition] with query pattern as argument.
     *
     * @param pattern a pattern to match against.
     * @param knowns known resolutions.
     *
     * @return updated knowns if the pattern matches, or `null` otherwise.
     */
    fun match(pattern: RulePattern, knowns: Knowns): Knowns? {
        if (pattern.terms.size != terms.size) return null

        var result = knowns.startMatching()
        var index = 0

        @Suppress("UseWithIndex")
        for (term in terms) {
            result = term.match(pattern.terms[index], result) ?: return null
            ++index
        }

        return result
    }

    /**
     * Creates a resolution rule with this pattern as its [condition][Rule.condition].
     *
     * @param predicate predicate the constructed rule resolves to if this pattern matches.
     */
    fun rule(predicate: Predicate) = Rule(this, predicate)

    /**
     * Creates a resolution rule with this pattern as its [condition][Rule.condition].
     *
     * @param predicate predicate the constructed rule resolves to if this pattern matches.
     */
    fun rule(predicate: PredicateFn) = Rule(this, predicate.asPredicate())

    /**
     * Creates a fact with this pattern as its [condition][Rule.condition].
     */
    fun fact() = rule(True)

    /**
     * Creates a rule resolved by [rules application][resolve].
     *
     * @param terms terms the rule search pattern consists of.
     */
    fun resolveBy(vararg terms: PlainTerm) = rule(RulePattern(*terms))

    /**
     * Creates a rule [resolved by phrase predicate][Phrase.resolve] consisting of the given terms.
     *
     * @param terms terms the phrase consists of.
     */
    fun resolveBy(vararg terms: Term) = rule(Phrase(*terms))

    /**
     * Searches for the rules matching this pattern and applies them.
     */
    override fun resolve(resolver: PredicateResolver): Flux<Knowns> =
            resolver.matchingRules(this@RulePattern, resolver.knowns)
                    .flatMap { (rule, knowns) ->
                        rule.predicate(resolver.withKnowns(knowns))
                    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as RulePattern

        if (terms != other.terms) return false

        return true
    }

    override fun hashCode(): Int {
        return terms.hashCode()
    }

    override fun toString() = printTerms(terms)

}
